pkg/sensors/accelerator/device/device.go

```diff
@@ -0,0 +1,156 @@
+	"sync"
+	"time"
+
+	"golang.org/x/exp/maps"
+	"k8s.io/klog/v2"
+)
+
+const (
+	DUMMY = iota
+	HABANA
+	DCGM
+	NVML
+)
+
+var (
+	globalRegistry *DeviceRegistry
+	once           sync.Once
+)
+
+type DeviceType int
+
+// Function prototype to create a new deviceCollector.
+type deviceStartupFunc func() DeviceInterface
+type DeviceRegistry struct {
+	gpuDevices   map[DeviceType]deviceStartupFunc // Static map of supported gpuDevices.
+	dummyDevices map[DeviceType]deviceStartupFunc // Static map of supported dummyDevices.
+}
+type DeviceInterface interface {
+	// Name returns the name of the device
+	Name() string
+	// DevType returns the type of the device (nvml, dcgm, habana ...)
+	DevType() DeviceType
+	// DevTypeName returns the type of the device (nvml, dcgm, habana ...) as a string
+	DevTypeName() string
+	// GetHwType returns the type of hw the device is (gpu, processor)
+	HwType() string
+	// InitLib the external library loading, if any.
+	InitLib() error
+	// Init initizalizes and start the metric device
+	Init() error
+	// Shutdown stops the metric device
+	Shutdown() bool
+	// DevicesByID returns a map with devices identifying then by id
+	DevicesByID() map[int]any
+	// DevicesByName returns a map with devices identifying then by name
+	DevicesByName() map[string]any
+	// DeviceInstances returns a map with instances of each Device
+	DeviceInstances() map[int]map[int]any
+	// AbsEnergyFromDevice returns a map with mJ in each gpu device. Absolute energy is the sum of Idle + Dynamic energy.
+	AbsEnergyFromDevice() []uint32
+	// DeviceUtilizationStats returns a map with any additional device stats.
+	DeviceUtilizationStats(dev any) (map[any]any, error)
+	// ProcessResourceUtilizationPerDevice returns a map of UtilizationSample where the key is the process pid
+	ProcessResourceUtilizationPerDevice(dev any, since time.Duration) (map[uint32]any, error)
+	// IsDeviceCollectionSupported returns if it is possible to use this device
+	IsDeviceCollectionSupported() bool
+	// SetDeviceCollectionSupported manually set if it is possible to use this device. This is for testing purpose only.
+	SetDeviceCollectionSupported(bool)
+}
+
+func (d DeviceType) String() string {
+	return [...]string{"DUMMY", "HABANA", "DCGM", "NVML"}[d]
+}
+
+// Registry gets the default device DeviceRegistry instance
+func Registry() *DeviceRegistry {
+	once.Do(func() {
+		globalRegistry = &DeviceRegistry{
+			gpuDevices:   map[DeviceType]deviceStartupFunc{},
+			dummyDevices: map[DeviceType]deviceStartupFunc{},
+		}
+	})
+	return globalRegistry
+}
+
+// SetRegistry replaces the global registry instance
+// NOTE: All plugins will need to be manually registered
+// after this function is called.
+func SetRegistry(registry *DeviceRegistry) {
+	globalRegistry = registry
+}
+
+// AddDeviceInterface adds a supported device interface, prints a fatal error in case of double registration.
+func AddDeviceInterface(dtype DeviceType, accType string, deviceStartup deviceStartupFunc) {
```

Comment:
Lets look at the coupling here ...

All devices (currently those under `accelerator/sources`) folder have to be
registered here.

This requires calling (AddDeviceInterface aka `Register`) with the following information
* type of device which is `DUMMY` , `HABANA`, `DCGM`, `NVML`
* further more it needs to specify the type of accelerator DUMMY, GPU

So the devices registry knows about all the devices it support and also that there is an accel type

However, this function also already hardcodes information about DCGM and knows what AccelType each device type is . 
So I would argue that  AccelType need not be specified by the caller. 

Further more adding or deleting a new device requires us to change the
DeviceType enum (coupling).

Next the point of this function is to register the device to the global registry, so the function needs to be part of registry since
the package that is trying to register. 
Changing this to a convenience function is fine though.
```
func Register(dtype Type, accType string, deviceStartup deviceStartupFunc) {
	globalRegistry.Register(dtype, accType, deviceStartup)
}
```

Regarding the call to Register, A Device that is registering already knows its
type and how to start, so shouldn't this  be a method of the device itself. In short, shouldn't dgcm look as follows

```
import "pkg/sensors/accelerator/devices"

type dcgm struct {
}

func (d *dcgm) Type() device.Type {
	return device.DCGM
}

func (d *dcgm) Start() error  {
	...
}

func Init() {
  devices.Register(&dcgm{})
}
```